app/server/apiGatewayHandler.ts

```diff
@@ -0,0 +1,160 @@
+) => resource.ResourceType === resourceTypeFilter;
+
+const toDefinedPhysicalResourceId = (stackName: string) => (
+  resource: StackResourceSummary
+) => {
+  if (resource.PhysicalResourceId) {
+    return resource.PhysicalResourceId;
+  }
+  throw new Error(
+    `PhysicalResourceId missing for '${resource.ResourceType}' of ${stackName}`
+  );
+};
+
+const getHost = (
+  stackName: string,
+  stage: string,
+  stackResourceSummaries?: StackResourceSummaries
+) => {
+  const hosts = stackResourceSummaries
+    ?.filter(byResourceType("AWS::ApiGateway::RestApi"))
+    .map(toDefinedPhysicalResourceId(stackName))
+    .map(
+      apiGatewayId =>
+        `${apiGatewayId}.execute-api.${AWS_REGION}.amazonaws.com/${stage}`
+    );
+  if (hosts && hosts.length === 1) {
+    return hosts[0];
+  }
+  log.error(`${(hosts || []).length} hosts for ${stackName}, expected 1`);
+};
+
+const lookupApiKey = async (apiKey: string) =>
+  (
+    await APIGateway.getApiKey({
+      apiKey,
+      includeValue: true
+    }).promise()
+  ).value;
+
+const getApiKeyPromise = (
+  stackName: string,
+  stage: string,
+  stackResourceSummaries?: StackResourceSummaries
+) => {
+  const apiKeyPromises = stackResourceSummaries
+    ?.filter(byResourceType("AWS::ApiGateway::ApiKey"))
+    .map(toDefinedPhysicalResourceId(stackName))
+    .map(lookupApiKey);
+  if (apiKeyPromises && apiKeyPromises.length === 1) {
+    return apiKeyPromises[0];
+  }
+  log.error(
+    `${(apiKeyPromises || []).length} API keys for ${stackName}, expected 1`
+  );
+};
+
+const getHostAndApiKeyForStack = (apiName: ApiName, stage: string) => {
+  const stackName = `membership-${stage}-${apiName}`;
+  return CloudFormation.listStackResources({
+    StackName: stackName
+    // no resources in question have anywhere near enough resources per-stack to require pagination with 'NextToken'
+    // https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_ListStackResources.html
+  })
+    .promise()
+    .then(async result => ({
+      host: getHost(stackName, stage, result.StackResourceSummaries),
+      apiKey: await getApiKeyPromise(
+        stackName,
+        stage,
+        result.StackResourceSummaries
+      )
+    }))
+    .catch(log.error);
+};
+
+export type PathnameTransformer = (pathParams: any) => string;
+
+export const apiGatewayHandler = (
+  apiName: ApiName,
+  pathnameTransformer: PathnameTransformer
+) => {
+  const normalModeConfigPromise = getHostAndApiKeyForStack(
+    apiName,
+    normalUserApiStage
+  );
+  const testModeConfigPromise = getHostAndApiKeyForStack(
+    apiName,
+    testUserApiStage
+  );
+
+  return async (req: express.Request, res: express.Response) => {
+    const isTestUser = req.header(MDA_TEST_USER_HEADER) === "true";
+    const { host, apiKey } = (await (isTestUser
+      ? testModeConfigPromise
+      : normalModeConfigPromise)) as any;
+    const stage = isTestUser ? testUserApiStage : normalUserApiStage;
+    if (!apiKey) {
```

Comment:
Ah I see why, it depends on null for error handling.

Reply:
yes we serve 500 for these services if we dont have the host or api key to call them, we'd then catch these with alarms.